feat(MOPLAT-331): dark mode

pvinis opened the pull request:
<!-- 
➡️ PR title in form of `type(PROJECT-XXXX): what changed`
➡️ Jira ticket in square brackets like [PROJECT-XXXX]
❗️ If this is a work in progress, remember to prefix it with [WIP] and/or open a draft PR instead of normal PR
-->

This PR resolves [MOPLAT-331]

### Description

dark mode 😎 ⚫ 

Current state

https://user-images.githubusercontent.com/100233/180281660-ba3ee394-6f3f-4fd2-bce9-a875cf0365af.mov



<!-- Info, implementation, how to get there, before & after screenshots & videos, follow-up work, etc -->

### PR Checklist

- [ ] I tested my changes on **iOS** and **Android**.
- [ ] I added screenshots or videos to illustrate my changes.
- [ ] I added Tests and Stories for my changes.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to run this PR on the simulator or device.

Need help with something? Have a look at our [docs], or get in touch with us.

[docs]: /docs/README.md


[MOPLAT-331]: https://artsyproduct.atlassian.net/browse/MOPLAT-331?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

pvinis commented on src/app/App.tsx:
```diff
@@ -1,30 +1,9 @@
 
 LogBox.ignoreLogs(["Expected style "])
 
-const AppProviders = ({ children }: { children: ReactNode }) => (
```
moved these in a separate file

pvinis commented on src/app/AppProviders.tsx:
```diff
@@ -0,0 +1,41 @@
+  }, [])
+  const isDarkMode = GlobalStore.useAppState((state) => state.devicePrefs.colorScheme)
+
+  return (
```
above is the trigger for dark mode and the calculation for the prop below.

pvinis commented on src/app/routes/HomeTabsNavigationStack.tsx:
```diff
@@ -26,30 +30,36 @@ export type HomeTabsScreens = {
+  const isDarkMode = GlobalStore.useAppState((s) => s.devicePrefs.colorScheme === "dark")
+  const color = useColor()
+  return (
+    <NavigationContainer theme={isDarkMode ? DarkTheme : DefaultTheme}>
```
here is the main change on this file. this is needed so that the screen transition on nav is not flashing white.

pvinis commented on src/app/screens/Auth/Login.tsx:
```diff
@@ -179,7 +179,7 @@ export const LoginScreenContent: React.FC = () => {
         </Text>
       </ScrollView>
-      <Flex px={2} paddingBottom={space(2)}>
+      <Flex px={2} paddingBottom={insets.bottom !== 0 ? insets.bottom : 20}>
```
tiny fix for the login screen, the logo on the bottom was WAY to low.

pvinis commented on src/app/screens/HomeTabs/Albums/CreateOrEditAlbum/CreateOrEditAlbum.tsx:
```diff
@@ -123,7 +123,7 @@ export const CreateOrEditAlbum = () => {
           <Flex flexDirection="row" alignItems="center" justifyContent="space-between">
             <Text>Add Items to Album</Text>
-            <ArrowRightIcon fill="black100" />
+            <ArrowRightIcon fill="contrast" />
```
`contrast` is the color that contrasts with the background. so white for dark mode, and black for light mode.

pvinis commented on src/app/screens/HomeTabs/Albums/CreateOrEditAlbum/CreateOrEditAlbumChooseArtworks.tsx:
```diff
@@ -76,13 +76,9 @@ export const CreateOrEditAlbumChooseArtworks = () => {
-              </Text>
-            </Touchable>
-          </Flex>
+          <Button size="small" onPress={() => selectAllArtworkHandler(!areAllArtworkSelected)}>
```
just using a `Button` instead

pvinis commented on src/app/screens/HomeTabs/HomeTabs.tsx:
```diff
@@ -21,7 +24,18 @@ export const HomeTabs = () => {
+        <Header
+          {...props}
+          withoutBackButton
+          leftElements={
```
new prop, for left buttons

pvinis commented on src/app/screens/Settings/DarkModeSettings.tsx:
```diff
@@ -0,0 +1,44 @@
+import { Flex, Screen, Text } from "palette"
+import { Switch } from "react-native"
+
+export function DarkModeSettings() {
```
a janky screen lol

pvinis commented on src/app/sharedUI/Header.tsx:
```diff
@@ -1,33 +1,43 @@
-    <Flex px={2}>
-      {!withoutBackButton ? (
-        <Flex mt={2}>
+    <Flex px={2} mt={2} flexDirection="row" alignItems="center">
```
a bit of layout fixes to support left elements.

pvinis commented on src/app/sharedUI/items/ShowListItem.tsx:
```diff
@@ -16,10 +16,10 @@ export const ShowListItem: React.FC<ShowListItemProps> = (props) => {
         {show?.name}
       </Text>
-      <Text variant="xs" color="black60">
+      <Text variant="xs" color="textFaded">
```
look at the `Theme.tsx` file for more info about these.

pvinis commented on src/app/store/Models/DevicePrefsModel.ts:
```diff
@@ -0,0 +1,37 @@
+}
+
+export const getDevicePrefsModel = (): DevicePrefsModel => ({
+  colorScheme: computed([(_, store) => store], (store) =>
```
here is the main calculation. basically, if we sync with system, we follow what the system has. if we don't sync with the system, then we fallback to the forced theme.

pvinis commented on src/palette/Theme.tsx:
```diff
@@ -31,8 +31,11 @@ export type Color =
+  | "contrast"
+  | "text"
+  | "textInactive"
+  | "textFaded"
```
here is the new layer of names. background and text are easy. contrast is the color that buttons get for example, the color that contrasts the most with bg. textInactive and textFaded are both colors that the text can get when it is not the main text of the screen.

Disclaimer: Names might change, when the design team starts incorporating dark mode in the design system.

pvinis commented on src/palette/Theme.tsx:
```diff
@@ -203,9 +211,12 @@ const THEMES = {
   v5dark: {
```
its called v5 for now, cause i think we will be done with the dark mode on DS before we arrive to v5. whenever we have dark mode on the DS, we can rename this v5 and v5dark to v4 or whatever version the DS has dark mode.

patrinoua replied:
we only have one version of dark mode right?

pvinis replied:
yea

pvinis commented on src/palette/elements/Text/Text.tsx:
```diff
@@ -30,6 +30,7 @@ export const Text = forwardRef(
     {
       variant = "sm",
       italic = false,
+      color = "text",
```
default text color is now text, instead of just black.

pvinis commented on src/palette/svgs/ArrowRightIcon.tsx:
```diff
@@ -1,14 +1,13 @@
       <Path
         d="M5.94 15.94l-.88-.88L11.12 9 5.06 2.94l.88-.88L12.88 9z"
-        fill={color(props.fill)}
+        fill={color(fill)}
```
not sure how our icons worked so far, but this makes `fill` not go to the `Icon` as a prop, which makes more sense. Also, it was complaining about some strings being there, so it feels cleaner now. I only changed the icons we actually use in energy. we can do more changes when we extract palette.

pvinis commented on src/palette/svgs/HamburgerIcon.tsx:
```diff
@@ -0,0 +1,35 @@
+  return (
+    <Icon {...restprops} viewBox="0 0 18 18">
+      <Path
+        d="
```
handcrafted svg like a boss. i enjoyed making it, and i laid it out in code in a way that we humans can actually read.

in case anyone is interested:
M: absolute move. so M 0,3 we move to x 0 and y 3, so we are on the left side of the square we will draw, and a bit down. (0,0 is top left).
H: absolute horizontal line. so H 12 draws a horizontal line from where we are (0,3 from above), to x 12. so we have now drawn a line from 0,3 to 12,3.
v: relative vertical line. so v 1 goes down 1 on the y axis. so we draw a line from 12,3 to 12,4
H 0: we draw a horizontal like to x 0, so from 12,4 to 0,4.
V 0: we draw a vertical like to y 0. (i realize this could have been `v -1` instead of `V 0` but ok.)
Z: ends the path.
So up to that point, we have drawn the top rect of the hamburger icon. Repeat two times with a small `m`, so the move is relative from the previous one. so by moving the first M 0,3 to another x and y, we move the whole hamburger.

patrinoua replied:
We already have an icon for that in palette it's called **MenuIcon**. So I am removing this

patrinoua reviewed (changes requested):
Kudos on these changes! Just one thing I'd like to point out:

I think we should not be touching the colors that we already have. 
We can just create translation/mapping of colors, as it is also on this [Dark Mode Figma file](https://www.figma.com/file/9CQuiUjVClr304DDjOI5UX/Hackathon%3A-Dark-Mode?node-id=0%3A1).

By creating a translation we can be sure that every color has a 1-1 correspondence to its dark-theme sibling.

The current implementation with "background", "contrast" and "text" does not offer any information about the actual color, plus we'd need to go for background100, background60, background30, text100, text60, text30, etc, which could be very confusing if we really end up using the dark theme everywhere.

The solution could be as simple as creating a function is inside useColor that does:

white100 => black100
black60 => black30
black30 => black60

green100=>red100

etc.

pvinis commented:
We tried this when we paired, and it seemed to work, but when working on it more it definitely seemed like a thing that would break sooner or later, for specific cases when we need the actual color without the mapping. So after some more thought I went with the original trial of the extra naming layer that seems more futureproof.

patrinoua commented:
In which way did it seem like it would break sooner or later? 
If we need a specific color we could do this locally, but it definitely feels wrong replacing all our colors with "background", "text", "contrast" etc. :/

patrinoua commented:
I'm reading this to get some more opinions on the topic 

https://blog.zenhub.com/improving-design-system-for-dark-mode-theming/

https://www.google.com/search?q=how+to+code+dark+mode+on+your+design+system&oq=how+to+code+dark+mode+on+your+design+system&aqs=chrome..69i57j0i22i30j0i390l3j69i64l3.24864j0j7&sourceid=chrome&ie=UTF-8

pvinis commented:
felt like it would break when we need a raw color without the mapping. we even had a hard time for the dark-mode-only avatar outline. and adding a manual `"white"` there is not making things easier. All colors should come from the tokens/DS. It makes things simple to have hex strings like `#123456`, name these like `orange70`, and then name our usages like `background` or `buttonDisabled`.

We are doing the same thing, but the 2-layer approach allows the designers to assign the same name color to many usage colors, and if they change one, all of them change. It also allows devs to handle colors better, depending if we want the `background` color or if we actually want the `black100` color, based on if its a usage as background or similar, or if we actually just need the color black100, or even specifically the darkest color of the display, which in that case we could use #000. Having two name layers makes things more easy to work with. One layer is great for one theme. Two layers help us go from one theme to two, but honestly if we wanted fifteen themes, we could easily make them using just the 2 named layers.


Also, your first link suggests exactly the same, with the `primary`, `secondary` names, that point to `black-60` and `black-0` for light and dark. Many others links and people suggest the same, and I have found this to be more convenient than just a mapping that bluntly replaces a hex string for a raw color name like black100 from #000 to #fff.

rajsam003 commented:
Great work @pvinis 👏
I just noticed a few things while running this branch locally that need to be fixed

1. Loading spinner screen which is white. 


https://user-images.githubusercontent.com/17421923/181197429-ab2385ff-9594-43ba-b928-4f99de393c07.mov






2. Buttons with the active color blue looks weird. 

https://user-images.githubusercontent.com/17421923/181197669-286d3ece-bb26-4319-be68-6774cf2b5e74.mov



3. Text input not visible

https://user-images.githubusercontent.com/17421923/181197901-43a2da96-cedf-4a9e-9496-b3d9f21d00b5.mov


4. `Currently available in all albums` text alignment is messed up

![Screenshot 2022-07-27 at 10 03 24](https://user-images.githubusercontent.com/17421923/181198109-68c7f9eb-8472-46bc-abdc-693cc19b1f37.png)

pvinis commented on src/app/navigation/HomeTabsNavigationStack.tsx:
```diff
@@ -32,27 +36,141 @@ export type HomeTabsScreens = {
+  )
+}
+
+const ThemeChoices = () => {
```
this will be removed before we deliver this app to users. its temporarily here for genia to play around with.

patrinoua commented:
We already have a Settings screen, let's maybe rebase to use it? 

https://github.com/artsy/energy/pull/120

src/app/screens/HomeTabs/Settings/Settings.tsx

pvinis commented:
is it already rebased, and i reworked them a bit to start using Screen and Screen.Body etc. I also made a SettingsItem and moved it in the sharedUI dir, so we can make settings items in the same way.

patrinoua commented:
I feel like changing our color scheme without consulting designers first is a change that is going to create an unnecessary cognitive burden and I personally am very much against doing this right now.

Since this is not a pressing matter, but rather a nice thing to have, instead of doing this change I would feel much better if we waited until the designers have decided on a color scheme and naming before changing anything.

Having said that, if this is something the whole team is onboard with, feel free to go ahead and merge.

pvinis commented:
i have consulted with the design team, this is the temp thing we will use. this PR has all the work to support dark mode, and the final colors are the thing we are waiting on.

i can make it so that the toggle just doesn't flip, so it is not visible to the user, if you think this is better. thought i don't see how this changes anything, since we are still not releasing this as is, and its useful for genia to play around with it while figuring out the final colors.

pvinis commented:
for completeness, this is the ticket explaining what we need to change once the design team has picked the final colors: https://artsyproduct.atlassian.net/browse/MOPLAT-381, very simple changes.

also, we do have a figma file, its here https://www.figma.com/file/BQVn5aOJ4varTAgtz8mnai/Folio-2022?node-id=1493%3A6887, and its a WIP.

ill see if i can separate things somehow, but i don't think i can 🤔

patrinoua commented on src/app/sharedUI/screens/Artwork/ArtworkContent/ArtworkContent.tsx:
```diff
@@ -79,21 +79,26 @@ export const ArtworkContent = ({ slug }: { slug: string }) => {
+              height={30}
+              alignItems="center"
+              justifyContent="center"
+              backgroundColor="background"
```
Ideally Flex has "background" as a background color by default?

pvinis replied:
I had that initially, but turns out that's a bad default, because that would make aaaaaal flexes need to have a color for a lot of places.

patrinoua commented on src/app/sharedUI/screens/Artwork/ArtworkContent/ArtworkContent.tsx:
```diff
@@ -79,21 +79,26 @@ export const ArtworkContent = ({ slug }: { slug: string }) => {
       >
-        <ScrollView style={{ paddingHorizontal: space(2) }} scrollEnabled={isScrollEnabled}>
+        <ScrollView
+          style={{ paddingHorizontal: space(2), backgroundColor: color("background") }}
```
Ideally ScrollView has "background" as a background color by default?

pvinis replied:
same as above. I had the same intuition. turns out, it's a tiring default.

patrinoua commented on src/app/sharedUI/screens/Artwork/ArtworkContent/ArtworkContent.tsx:
```diff
@@ -102,11 +107,11 @@ export const ArtworkContent = ({ slug }: { slug: string }) => {
             </>
           )}
-          <Text variant="xs" color="black60">
+          <Text variant="xs" color="onBackgroundMedium">
```
Ideally Text variant="xs" has "onBackgroundMedium" as color by default?

pvinis replied:
ooo interesting. maybe? 🤔 thought I like to keep it simple for defaults, so all text the same? could be that we should do that in the future

patrinoua commented on src/app/sharedUI/screens/Artwork/ArtworkContent/ArtworkContent.tsx:
```diff
@@ -132,14 +137,14 @@ export const ArtworkContent = ({ slug }: { slug: string }) => {
               <Spacer mt={3} />
               <Flex flexDirection="row" alignItems="center">
-                <BriefcaseIcon fill="black100" />
+                <BriefcaseIcon fill="onBackgroundHigh" />
```
Ideally all Icons have "onBackgroundHigh" as color by default?

pvinis replied:
good point here. I saw that eigen has half the icons on background and half on primary on buttons. so I wasn't sure what to default to, and left it as it was already 🤷‍♂️

patrinoua commented on src/palette/atoms/BackButton/BackButton.tsx:
```diff
@@ -6,35 +6,31 @@ interface BackButtonProps {
+export const BackButtonWithBackground: React.FC<BackButtonProps> = ({ onPress, showX = false }) => (
+  <TouchableOpacity onPress={onPress}>
+    <Flex
+      backgroundColor="white100"
```
this should be "background" right?

pvinis replied:
I did not touch that button because we don't we it on energy I think? if we do, maybe we need to change that yea.

patrinoua commented on src/palette/elements/Avatar/Avatar.tsx:
```diff
@@ -48,6 +50,8 @@ export const Avatar = ({ src, initials, size = DEFAULT_SIZE }: AvatarProps) => {
           width: diameter,
           height: diameter,
           borderRadius: diameter / 2,
+          borderColor: color("white100"),
```
this should use a color from our new colorscheme, right?

pvinis replied:
this one no. this is always white. weade it together, remember? this is always white because its the outline. invisible on light mode, and contrast-y on dark mode.

ideally we should have the "outline" color that we can set to "transparent" on light and "white100" on dark. 🤔 but I don't know if we have outline elsewhere.

I guess a good candidate together with the traffic light colors 👌

patrinoua commented on src/palette/elements/Input/InputTitle.tsx:
```diff
@@ -1,41 +1,29 @@
+    <Text variant="md" style={{ fontSize: 13, marginBottom: 2, textTransform: "uppercase" }}>
+      {title}
+      {!!required && (
+        <Text variant="md" style={{ fontSize: 13, textTransform: "none" }} color="black60">
```
onBackgroundMedium

pvinis replied:
true 👍 I missed that

patrinoua commented on src/palette/elements/Input/InputTitle.tsx:
```diff
@@ -1,41 +1,29 @@
+        </Text>
+      )}
+      {!!optional && (
+        <Text variant="md" style={{ fontSize: 13, textTransform: "none" }} color="black60">
```
onBackgroundMedium

pvinis replied:
same. if you can, make a PR for these two, or I'll do it later.

patrinoua replied:
I've made a PR here for approaching some of the things introduced by this PR but also address some other thigns.
Please have a look :) https://github.com/artsy/energy/pull/136

patrinoua replied:
There are some initial changes there, I plan to do more changes like adding colors on Monday. This is more infra work so far. so if possible let's merge and I'll open a new one for fixing other issues
